pkg/apb/bind.go

```diff
@@ -37,41 +37,55 @@ func Bind(instance *ServiceInstance,
 	log.Notice(fmt.Sprintf("ServiceInstance.Description: %s", instance.Spec.Description))
 	log.Notice("============================================================")
 
-	executionContext, err := ExecuteApb("bind", instance.Spec, instance.Context, parameters)
-	defer runtime.Provider.DestroySandbox(
-		executionContext.PodName,
-		executionContext.Namespace,
-		executionContext.Targets,
-		clusterConfig.Namespace,
-		clusterConfig.KeepNamespace,
-		clusterConfig.KeepNamespaceOnError,
-	)
-	if err != nil {
-		log.Errorf("Problem executing apb [%s] bind", executionContext.PodName)
-		return executionContext.PodName, nil, err
-	}
-	k8scli, err := clients.Kubernetes()
-	if err != nil {
-		log.Error("Something went wrong getting kubernetes client")
-		return executionContext.PodName, nil, err
-	}
+	go func() {
+		e.start()
+		metrics.ActionStarted("bind")
+		executionContext, err := e.executeApb(
+			"bind", instance.Spec, instance.Context, parameters)
+		defer runtime.Provider.DestroySandbox(
```

Comment:
I'm not clear on the best practice with defers. If I'm doing something and I know I'll need to call the relevant "finish" function at the end of it, it makes sense to me to defer the call immediately after, assuming the mechanics of defer apply. A good example of this is the metrics hooks in the `apbJob`, I just partner the Start with a defered finish right way. I'll take a closer look at this, thanks for raising it.

Reply:
After looking at this, I think it doesn't actually matter in practice? Regardless of where the defer call is made, it will execute at the conclusion of the function. Optically, it might make sense to be made elsewhere, though.

Reply:
It doesn't matter, what I was saying was tying statements together.

```golang
    go func() {
        e.start()
        metrics.ActionStarted("bind")

        defer runtime.Provider.DestroySandbox(
            executionContext.PodName,
            executionContext.Namespace,
            executionContext.Targets,
            clusterConfig.Namespace,
            clusterConfig.KeepNamespace,
            clusterConfig.KeepNamespaceOnError,
        )

        executionContext, err := e.executeApb("bind", instance.Spec, instance.Context, parameters)
        if err != nil {
            log.Errorf("Problem executing apb [%s] bind", executionContext.PodName)
            e.finishWithError(err)
            return
        }
```

So the `err != nil` follows the executeApb so it is obvious that it is related and highly unlikely that someone will come and add another line of code that set err. As it stands now the defer statement separates the executeApb and the err check. If someone comes in later they could accidentally set err and we'll be eating an err.

Reply:
The issue is that `executionContext` isn't in scope until the `executeApb` call finished. I'll update to keep the err check tied closely.